### PR TITLE
Extract service logic from controllers

### PR DESCRIPTION
**Pull request by ThanGerlek**
Converts the "endpoint" Controllers (i.e. excluding `TrafficController` and `WebSocketController`) into the beginnings of a Handler layer, and moves most of the logic into Service classes.

This should improve testability and compliance with SRP, Demeter, and DRY. It also isolates and simplifies JavaSpark code, which will make it much easier to migrate to Javalin.

### CODE REVIEW

**Review by frozenfrank — CHANGES_REQUESTED** [review]
It looks pretty good to me. I'm really happy that we are properly dividing the app logic and separating concerns according to the good principles were learning from our professors. Thank you @jerodw!

There are some improvements I would still make as discussed in the comments. Additional code duplication to pull out, and a few organizational choices.

File: src/main/java/edu/byu/cs/service/AdminService.java [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    }
+
+    public static void updateUser(User user) throws DataAccessException, ItemNotFoundException {
+        UserDao userDao = DaoService.getUserDao();
+        User existingUser;
+        try {
+            existingUser = userDao.getUser(user.netId());
+        } catch (DataAccessException e) {
+            LOGGER.error("Error getting user", e);
+            throw e;
```
[the remark this answers is not in the document]
Comment (ThanGerlek) [reply src/main/java/edu/byu/cs/service/AdminService.java:47]: Agreed, but not in this PR--I'm working on an overhaul of all the Handler/Service exception handling. This PR is focused on extracting the Service layer while preserving logical structure.

[the remark this answers is not in the document]
Comment (frozenfrank) [reply src/main/java/edu/byu/cs/service/AdminService.java:47]: I'm beginning to detect that we have different attitudes about the role of a PR. Again, these are kinds of discussions are stylistic/conventional measures; I don't want to say either of us are right, but it will help us collaborate better together if we can assimilate and be in the same page.

It seems like there are two different perspectives that can be taken towards a PR:
* A PR is like a commit: one comes now and another will change it later
* A PR is like a branch: many commits will be added to each other until the meaningful milestone or objective is complete

**TL;DR** PRs serve as an organizational unit of changes that achieve a business goal. Use commits to express steps of the work, but keep related changes contained within the same PR.


I personally believe a PR should be more like a branch-- it can contain many commits, potentially including many that individually touch lots of files, and in the end, the PR will correlate closely to an actual business result. It may be that I discover halfway through a PR, that I receive new information which invalidates some of the assumptions made in the code, and so I would either build on the PR to fix it, or I may restart with a new PR to try a new approach.

The aligns with my experience where certain business needs are the end goal, and multiple people collaborate together on a PR to make sure that the result is the best possible for the company.

In my experience, the purpose of a PR is to represent something to the business (not typically a mere minor technical improvement). A PR, or a a small handful of PRs, will be what I show my boss to show him what I did.

In certain cases, for the sake of urgency or sheer size, issues should be split into smaller items which would result in parallel reductions in the overall size of a corresponding PR. However, there is no large requirement here and no urgency to merge in the portion of code completed before we totally finish it.

The requested changes can be represented in a commit or two, and I think it would be totally appropriate to put them together under the banner of "Refactoring and Extracting Controller Code into the proper Locations." Breaking down the work into meaningful chucks at the commit level makes sense, but that level of detail should not directly equate to a linear number of PRs as well.

[the remark this answers is not in the document]
Comment (ThanGerlek) [reply src/main/java/edu/byu/cs/service/AdminService.java:47]: All right, what do you think is the best way to do that here then? I could introduce an intermediate refactoring to address just the requested changes, but it'll get completely replaced in the second PR anyway; or I could compile that entire PR into this one, but that will be a lot more than just a couple of commits--it would involve adding several new Exception classes and things of that ilk.

File: src/main/java/edu/byu/cs/service/UserService.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+        }
+
+        if (adminNetId == null) {
+            LOGGER.info("student {} changed their repoUrl to {}", studentNetId, repoUrl);
+        } else {
+            LOGGER.info("admin {} changed the repoUrl for student {} to {}", adminNetId, studentNetId, repoUrl);
+        }
+    }
+
+    private static boolean isValidRepoUrl(String url) {
```
[the remark this answers is not in the document]
Comment (frozenfrank) [reply src/main/java/edu/byu/cs/service/UserService.java:95]: I found the code that performs this work in `GitHelper`! We may have to expose the private function, but let's prefer to reuse code rather than allowing it to be duplicated.

File: src/main/java/edu/byu/cs/service/SubmissionService.java [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+
+    public static Collection<Submission> getSubmissionsForUser(String netId) throws DataAccessException {
+        return getXSubmissionsForUser(netId, null);
+    }
+
+    public static void approveSubmission(String adminNetId, ApprovalRequest request) throws GradingException, DataAccessException {
+        int penalty = 0;
+        if (request.penalize()) {
+            //TODO: Put somewhere better/more configurable
+            penalty = 10;
```
[the remark this answers is not in the document]
Comment (webecke) [reply src/main/java/edu/byu/cs/service/SubmissionService.java:188]: Not yet, config enhancements is next up on my TODO after the the Contributing Guide and ServerCommunicator are pushed through  (#448 and #419)

File: src/main/java/edu/byu/cs/service/AdminService.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (webecke) [reply src/main/java/edu/byu/cs/service/AdminService.java:47]: @frozenfrank I'm not sure a PR needs to be so broad though. Extracting service logic from controllers does seem like a big change, and other overhauls might be good to do separately.

File: src/main/java/edu/byu/cs/service/SubmissionService.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (ThanGerlek) [reply src/main/java/edu/byu/cs/service/SubmissionService.java:188]: I'll leave it as it is for now then.

File: src/main/java/edu/byu/cs/service/UserService.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (ThanGerlek) [reply src/main/java/edu/byu/cs/service/UserService.java:95]: I agree it shouldn't be here. That would introduce a secondary entry point and usage context for GitHelper, though, and at present GitHelper is very nicely self-contained and cohesive. Rather than broaden the Single Responsibility of GitHelper, I propose we rename GitHelper to GitGradingHelper or something similar, then create a GitHelper class containing pure static functions that don't depend on a GradingContext. Then both UserService and GitGradingHelper can depend on it.

Objections?

File: src/main/java/edu/byu/cs/service/SubmissionService.java [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+     *
+     * @param netId the netId of the user
+     * @param phase the phase to grade
+     * @param adminSubmission if the grader should run in admin mode
+     * @return the grader
+     * @throws IOException if there is an error creating the grader
+     */
+    private static Grader getGrader(String netId, Phase phase, String repoUrl, boolean adminSubmission) throws IOException, GradingException {
+        // TODO? Make into an actual class...?
+        GradingObserver observer = new GradingObserver() {
```
[the remark this answers is not in the document]
Comment (19mdavenport) [reply src/main/java/edu/byu/cs/service/SubmissionService.java:205]: Yeah making this a top level class sounds good. I think it was originally very small and has mutated since then

File: src/main/java/edu/byu/cs/service/AdminService.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (frozenfrank) [reply src/main/java/edu/byu/cs/service/AdminService.java:47]: @webecke Fair point. Let's wrap up this one, and represent related changes in other issues.

**Review by 19mdavenport — CHANGES_REQUESTED** [review]
From an in-person conversation it sounded like you were planning on doing this but I would just like to go officially on the record as requesting the GradingObserver implementation become its own class

**Comment by ThanGerlek** [comment]
How's this?

File: src/main/java/edu/byu/cs/service/UserService.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (ThanGerlek) [reply src/main/java/edu/byu/cs/service/UserService.java:95]: Moving conversation to #463

**Review by 19mdavenport — APPROVED** [review]

**Review by webecke — APPROVED** [review]
It looks really good seems to not have broken anything, and is something we probably should have done a long time ago. Good work!

**Comment by webecke** [comment]
Just checking in, @frozenfrank @ThanGerlek do we have an update on this PR?

**Comment by frozenfrank** [comment]
> Just checking in, @frozenfrank @ThanGerlek do we have an update on this PR?

@webecke Thanks for checking in. I just realized I still have a pending review comments I'll review it again now.

File: src/main/java/edu/byu/cs/service/CasService.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+import javax.net.ssl.HttpsURLConnection;
+import java.io.IOException;
+import java.net.URI;
+import java.util.Map;
+
+public class CasService {
+    private static final Logger LOGGER = LoggerFactory.getLogger(CasService.class);
+    public static final String BYU_CAS_URL = "https://cas.byu.edu/cas";
+
+    public static User callback(String ticket) throws InternalServerException, BadRequestException, DataAccessException, CanvasException {
```
[the remark this answers is not in the document]
Comment (frozenfrank) [reply src/main/java/edu/byu/cs/service/CasService.java:25]: This issue is now represented by #476.

**Review by frozenfrank — COMMENTED** [review]
I have a few adjustments that I've developed locally and will push the branch.

My main comment is an important comment, but I feel it will be appropriate to address in a separate issue. See #476.

**Review by frozenfrank — APPROVED** [review]
Let's move forward with this for now, and address the rest later.

Thank you for making these significant improvements to the AutoGrader!

**Comment by frozenfrank** [comment]
@ThanGerlek Are you planning on merging this PR?

On this team, we like to let the original author of the PR press the "Merge" button. Personally, I feel like the tradition leads to hiccups and additional delays, but it is how it is right now. (You may need to re-update the branch since other commits may have been added to `main` since you last updated it.)

**Comment by ThanGerlek** [comment]
I was, but I'm complete swamped, so I haven't had time to review your changes or the merge from main that still needs to happen. If you want to ping Dallin or Michael to review your changes, they or you can merge with main and then merge in the PR

**Comment by ThanGerlek** [comment]
Honestly, if you're confident that your changes are small enough, I trust your judgement if you want to go ahead and merge without review. Just be careful obviously

**Comment by frozenfrank** [comment]
> I was, but I'm complete swamped, so I haven't had time to review your changes or the merge from main that still needs to happen. If you want to ping Dallin or Michael to review your changes, they or you can merge with main and then merge in the PR

There's no rush, I just saw that you had updated the branch, and I expected the branch to be merged at the same time. Just review it when you have a moment 👌
